omod/src/main/java/org/openmrs/module/pihcore/fragment/controller/dashboardwidgets/DiscreteValuesFragmentController.java

```diff
@@ -0,0 +1,143 @@
+import org.openmrs.Program;
+import org.openmrs.api.ConceptService;
+import org.openmrs.api.ObsService;
+import org.openmrs.api.ProgramWorkflowService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.appframework.domain.AppDescriptor;
+import org.openmrs.module.emrapi.patient.PatientDomainWrapper;
+import org.openmrs.ui.framework.SimpleObject;
+import org.openmrs.ui.framework.annotation.FragmentParam;
+import org.openmrs.ui.framework.annotation.SpringBean;
+import org.openmrs.ui.framework.fragment.FragmentConfiguration;
+import org.openmrs.ui.framework.fragment.FragmentModel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DiscreteValuesFragmentController {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    public void controller(
+                           @SpringBean("conceptService") ConceptService conceptService,
+                           @SpringBean("obsService") ObsService obsService,
+                           @SpringBean("programWorkflowService") ProgramWorkflowService programWorkflowService,
+                           @FragmentParam("app") AppDescriptor app,
+                           FragmentConfiguration config,
+                           FragmentModel model) {
+
+        Object patientConfig = config.get("patient");
+        Patient patient = null;
+        if (patientConfig != null && (patientConfig instanceof PatientDomainWrapper)) {
+            patient = ((PatientDomainWrapper) patientConfig).getPatient();
+        }
+        if (patient == null) {
+            throw new IllegalArgumentException("No patient found. Please pass patient into the configuration");
+        }
+
+        model.put("app", app);
+        model.put("patient", patient);
+
+        JsonNode node = app.getConfig().get("fields");
+        if (node == null) {
+            throw new IllegalStateException("Missing configuration concepts on widget");
+        }
+        Set<Concept> concepts = new HashSet<>();
+        Map<String, SimpleObject> fields = new LinkedHashMap<>();
+        if (node instanceof ArrayNode) {
+            ArrayNode arrayNode = (ArrayNode) node;
+            for (Iterator<JsonNode> iter = arrayNode.getElements(); iter.hasNext();) {
+                JsonNode fieldNode = iter.next();
+                try {
+                    String label = fieldNode.get("label").asText();
+                    String conceptUuid = fieldNode.get("concept").asText();
+                    String minValue = fieldNode.get("minValue").asText();
+                    if (StringUtils.isNotBlank(conceptUuid)) {
+                        Concept concept = conceptService.getConceptByUuid(conceptUuid);
```

Comment:
Ideally this would support mappings equally well.  You could use `HtmlFormEntryUtil.getConcept(String lookup)` or the new method in core `conceptService.getConceptByReference(String ref)`, which supports uuid and a number of other lookup options.